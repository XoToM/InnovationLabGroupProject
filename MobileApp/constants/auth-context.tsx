"use client";

import type React from "react";
import { createContext, useState, useEffect, useContext } from "react";
import AsyncStorage from "@react-native-async-storage/async-storage";

// Define user type
type User = {
  id: string;
  email: string;
  name: string;
};

// Define auth context type
type AuthContextType = {
  user: User | null;
  loading: boolean;
  signIn: (email: string, password: string) => Promise<{ error?: string }>;
  signUp: (
    name: string,
    email: string,
    password: string
  ) => Promise<{ error?: string }>;
  signOut: () => Promise<void>;
};

// Create context
const AuthContext = createContext<AuthContextType | undefined>(undefined);

// Mock user database (in a real app, this would be your backend)
const MOCK_USERS = [
  {
    id: "1",
    email: "user@example.com",
    password: "password123",
    name: "Demo User",
  },
];

// Provider component
export const AuthProvider: React.FC<{ children: React.ReactNode }> = ({
  children,
}) => {
  const [user, setUser] = useState<User | null>(null);
  const [loading, setLoading] = useState(true);

  // Check for existing session on mount
  useEffect(() => {
    const loadUser = async () => {
      try {
        const userJson = await AsyncStorage.getItem("@auth_user");
        if (userJson) {
          setUser(JSON.parse(userJson));
        }
      } catch (error) {
        console.error("Failed to load user from storage", error);
      } finally {
        setLoading(false);
      }
    };

    loadUser();
  }, []);

  // Sign in function
  const signIn = async (email: string, password: string) => {
    try {
      // In a real app, this would be an API call
      const foundUser = MOCK_USERS.find(
        (u) => u.email === email && u.password === password
      );

      if (!foundUser) {
        return { error: "Invalid email or password" };
      }

      // Create user object without password
      const userToStore: User = {
        id: foundUser.id,
        email: foundUser.email,
        name: foundUser.name,
      };

      // Store user in state and AsyncStorage
      setUser(userToStore);
      await AsyncStorage.setItem("@auth_user", JSON.stringify(userToStore));

      return {};
    } catch (error) {
      console.error("Sign in error", error);
      return { error: "An unexpected error occurred" };
    }
  };

  // Sign up function
  const signUp = async (name: string, email: string, password: string) => {
    try {
      // Check if user already exists
      const userExists = MOCK_USERS.some((u) => u.email === email);

      if (userExists) {
        return { error: "User with this email already exists" };
      }

      // In a real app, this would create a user in your database
      // For this mock version, we'll just pretend it worked

      // You could add the user to MOCK_USERS here if you want to test signing in
      // with the newly created account

      return {};
    } catch (error) {
      console.error("Sign up error", error);
      return { error: "An unexpected error occurred" };
    }
  };

  // Sign out function
  const signOut = async () => {
    try {
      // Clear user from state and AsyncStorage
      setUser(null);
      await AsyncStorage.removeItem("@auth_user");
    } catch (error) {
      console.error("Sign out error", error);
    }
  };

  return (
    <AuthContext.Provider value={{ user, loading, signIn, signUp, signOut }}>
      {children}
    </AuthContext.Provider>
  );
};

// Custom hook to use auth context
export const useAuth = () => {
  const context = useContext(AuthContext);
  if (context === undefined) {
    throw new Error("useAuth must be used within an AuthProvider");
  }
  return context;
};
export { AuthContext };
