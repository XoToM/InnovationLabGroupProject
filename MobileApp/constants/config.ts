export const API_DOMAIN = 'https://katestudent.pythonanywhere.com'; //use for production testing
//export const API_DOMAIN = 'http://localhost:5000'; // use for local testing