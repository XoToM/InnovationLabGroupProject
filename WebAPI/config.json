{
    "db_config": {
        "host": "katestudent.mysql.pythonanywhere-services.com",
        "user": "katestudent",
        "password": "KateAdmin",
        "database": "katestudent$default"
    }
}