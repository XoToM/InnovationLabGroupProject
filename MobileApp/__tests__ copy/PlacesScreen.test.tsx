import React from "react";
import { TouchableOpacity } from "react-native";
import { render, waitFor, fireEvent } from "@testing-library/react-native";
import PlacesScreen from "../app/screens/places";
import { API_DOMAIN } from "@/constants/config";

// Mock data
const mockPlaces = [
  {
    idPlace: 1,
    name: "Library",
    formattedAddress: "123 Main St",
    latitude: 37.7749,
    longitude: -122.4194,
  },
  {
    idPlace: 2,
    name: "Park",
    formattedAddress: "456 Elm St",
    latitude: 40.7128,
    longitude: -74.006,
  },
];

// Mock fetch
global.fetch = jest.fn(() =>
  Promise.resolve({
    ok: true,
    json: () => Promise.resolve({ places: mockPlaces }),
  })
) as jest.Mock;

// Mock expo-router Link
jest.mock("expo-router", () => ({
  Link: ({ children, href }: any) => <>{children}</>,
}));

describe("PlacesScreen", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("renders loading state initially", () => {
    const { getByText } = render(<PlacesScreen />);
    expect(getByText("Loading...")).toBeTruthy();
  });

  it("fetches places and displays them", async () => {
    const { getByText } = render(<PlacesScreen />);
    await waitFor(() => {
      expect(getByText("Library")).toBeTruthy();
      expect(getByText("Park")).toBeTruthy();
    });
  });

  it("displays address and coordinates of each place", async () => {
    const { getByText } = render(<PlacesScreen />);
    await waitFor(() => {
      expect(getByText("123 Main St")).toBeTruthy();
      expect(getByText("456 Elm St")).toBeTruthy();
      expect(getByText("37.7749, -122.4194")).toBeTruthy();
      expect(getByText("40.7128, -74.006")).toBeTruthy();
    });
  });

  it("renders a FlatList of places with 2 columns", async () => {
    const { UNSAFE_getAllByType } = render(<PlacesScreen />);
    await waitFor(() => {
      const touchables = UNSAFE_getAllByType(TouchableOpacity);
      expect(touchables.length).toBeGreaterThanOrEqual(2);
    });
  });

  it("does not crash on empty list", async () => {
    (fetch as jest.Mock).mockResolvedValueOnce({
      ok: true,
      json: async () => ({ places: [] }),
    });

    const { queryByText } = render(<PlacesScreen />);
    await waitFor(() => {
      expect(queryByText("Library")).toBeNull();
    });
  });

  it("handles API error gracefully", async () => {
    (fetch as jest.Mock).mockResolvedValueOnce({
      ok: false,
    });

    const { getByText } = render(<PlacesScreen />);
    await waitFor(() => {
      expect(getByText("Loading...")).toBeTruthy(); // Still shows loading momentarily
    });
  });

  it("matches snapshot after data load", async () => {
    const { toJSON } = render(<PlacesScreen />);
    await waitFor(() => {
      const tree = toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  it("uses correct API endpoint", async () => {
    render(<PlacesScreen />);
    await waitFor(() => {
      expect(fetch).toHaveBeenCalledWith(`${API_DOMAIN}/places`);
    });
  });

  it("displays all place names correctly", async () => {
    const { getByText } = render(<PlacesScreen />);
    await waitFor(() => {
      mockPlaces.forEach((place) => {
        expect(getByText(place.name)).toBeTruthy();
      });
    });
  });

  it("renders the correct number of cards", async () => {
    const { UNSAFE_getAllByType } = render(<PlacesScreen />);
    await waitFor(() => {
      const cards = UNSAFE_getAllByType(TouchableOpacity);
      expect(cards.length).toBe(mockPlaces.length);
    });
  });
});
