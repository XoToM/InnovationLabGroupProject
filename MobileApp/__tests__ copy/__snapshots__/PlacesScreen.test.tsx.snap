// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlacesScreen matches snapshot after data load 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#f5f5a3",
      "flex": 1,
    }
  }
>
  <Text>
    Loading...
  </Text>
</View>
`;
